test/Models/TerriaSpec.ts

```diff
@@ -333,42 +310,357 @@ describe("Terria", function() {
+
+        await terria.start({
+          configUrl
+        });
+        jasmine.Ajax.stubRequest(configUrl).andReturn({
+          responseText: JSON.stringify(
+            require("../../wwwroot/test/Magda/shareKeys/map-config-example-new.json")
+          )
+        });
+
+        await newTerria.start({
+          configUrl
+        });
+        // Don't allow more requests to configUrl once Terrias are set up
+        jasmine.Ajax.stubRequest(configUrl).andError({});
+      });
+
+      afterEach(function() {
+        jasmine.Ajax.uninstall();
+      });
+
+      it("correctly applies user stratum changes to moved item", async function() {
+        const oldGroupRef = terria.getModelById(
+          MagdaReference,
+          "6b24aa39-1aa7-48d1-b6a6-9e755aff4476"
+        );
+        expect(oldGroupRef).toBeDefined(
+          "Can't find Old group reference in source terria"
+        );
+        if (oldGroupRef === undefined) return;
+        await oldGroupRef.loadReference();
+        expect(oldGroupRef.target).toBeDefined(
+          "Can't dereference Old group in source terria"
+        );
+
+        const csv = terria.getModelById(
+          CsvCatalogItem,
+          "3432284e-a111-4844-97c8-26a1767f9986"
+        );
+        expect(csv).toBeDefined("Can't dereference csv in source terria");
+        if (csv === undefined) return;
+        csv.setTrait(CommonStrata.user, "opacity", 0.5);
+        const shareLink = buildShareLink(terria, viewState);
+
+        // Hack to make below test succeed. This needs to be there until we add a magda API that can locate any
+        //  item by ID or share key within a Terria catalog
+        // Loads "New group" (bfc69476-1c85-4208-9046-4f736bab9b8e) which registers shareKeys for
+        //  "Extra group" (12f26f07-f39e-4753-979d-2de01af54bd1). And "Extra group" has a share key
+        //  that matches the ancestor of the serialised Random CSV, so loading is triggered on "Extra group"
+        //  followed by 3432284e-a111-4844-97c8-26a1767f9986 which points to "My random CSV"
+        //  (decfc787-0425-4175-a98c-a40db064feb3)
+        const newGroupRef = newTerria.getModelById(
+          MagdaReference,
+          "bfc69476-1c85-4208-9046-4f736bab9b8e"
+        );
+        if (newGroupRef === undefined) return;
+        await newGroupRef.loadReference();
+
+        await newTerria.updateApplicationUrl(shareLink);
+
+        // Why does this return a CSV item (when above hack isn't added)? It returns a brand new csv item without data or URL
+        // Does serialisation save enough attributes that upsertModelFromJson thinks it can create a new model?
+        // upsertModelFromJson should really be replaced with update + insert functions
+        // But is it always easy to work out when share data should use update and when it should insert?
+        // E.g. user added models should be inserted when deserialised, not updated
+        const newCsv = newTerria.getModelByIdOrShareKey(
```

Comment:
It might be worth making a ticket about this, as the error you get here is not great
![image](https://user-images.githubusercontent.com/6187649/100225223-f9947c80-2f71-11eb-810a-63a570339e39.png)

Reply:
I made one!
https://github.com/TerriaJS/terriajs/issues/5023